bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_collector_oda.cpp

```diff
@@ -0,0 +1,129 @@
+		}
+		OnTriangleOut(vertices);
+	}
+}
+
+double DataCollectorOda::deviation(
+	const OdGiDeviationType deviationType,
+	const OdGePoint3d& pointOnCurve) const {
+	return 0;
+}
+
+OdGiMaterialItemPtr DataCollectorOda::fillMaterialCache(
+	OdGiMaterialItemPtr prevCache,
+	OdDbStub* materialId,
+	const OdGiMaterialTraitsData & materialData
+) {
+	repo_material_t material;
+
+	OdGiMaterialColor diffuseColor; OdGiMaterialMap diffuseMap;
+	OdGiMaterialColor ambientColor;
+	OdGiMaterialColor specularColor; OdGiMaterialMap specularMap; double glossFactor;
+	OdGiMaterialColor emissiveColor; OdGiMaterialMap emissiveMap;
+
+	double opacityPercentage; OdGiMaterialMap opacityMap;
+
+	materialData.diffuse(diffuseColor, diffuseMap);
+	materialData.ambient(ambientColor);
+	materialData.specular(specularColor, specularMap, glossFactor);
+	materialData.opacity(opacityPercentage, opacityMap);
+	materialData.emission(emissiveColor, emissiveMap);
+
+	MaterialColors colors;
+
+	if (diffuseColor.color().colorMethod() == OdCmEntityColor::kByColor)
+	{
+		colors.colorDiffuse = ODTOCOLORREF(diffuseColor.color());
+	}
+	else if (diffuseColor.color().colorMethod() == OdCmEntityColor::kByACI)
+	{
+		colors.colorDiffuse = OdCmEntityColor::lookUpRGB((OdUInt8)diffuseColor.color().colorIndex());
+	}
+	if (ambientColor.color().colorMethod() == OdCmEntityColor::kByColor)
+	{
+		colors.colorAmbient = ODTOCOLORREF(ambientColor.color());
+	}
+	else if (ambientColor.color().colorMethod() == OdCmEntityColor::kByACI)
+	{
+		colors.colorAmbient = OdCmEntityColor::lookUpRGB((OdUInt8)ambientColor.color().colorIndex());
+	}
+	if (specularColor.color().colorMethod() == OdCmEntityColor::kByColor)
+	{
+		colors.colorSpecular = ODTOCOLORREF(specularColor.color());
+	}
+	else if (specularColor.color().colorMethod() == OdCmEntityColor::kByACI)
+	{
+		colors.colorSpecular = OdCmEntityColor::lookUpRGB((OdUInt8)specularColor.color().colorIndex());
+	}
+	if (emissiveColor.color().colorMethod() == OdCmEntityColor::kByColor)
+	{
+		colors.colorEmissive = ODTOCOLORREF(emissiveColor.color());
+	}
+	else if (emissiveColor.color().colorMethod() == OdCmEntityColor::kByACI)
+	{
+		colors.colorEmissive = OdCmEntityColor::lookUpRGB((OdUInt8)emissiveColor.color().colorIndex());
+	}
+	
+	colors.colorDiffuseOverride = diffuseColor.method() == OdGiMaterialColor::kOverride ? true : false;
+	colors.colorAmbientOverride = ambientColor.method() == OdGiMaterialColor::kOverride ? true : false;
+	colors.colorSpecularOverride = specularColor.method() == OdGiMaterialColor::kOverride ? true : false;
+	colors.colorEmissiveOverride = emissiveColor.method() == OdGiMaterialColor::kOverride ? true : false;
+
+	material.shininessStrength = glossFactor;
+	material.shininess = materialData.reflectivity();
+	material.opacity = opacityPercentage;
+
+	OnFillMaterialCache(prevCache, materialId, materialData, colors, material);
```

Comment:
Again avoid. Child should override this function instead of a listener style of abstract function.

If you want to do a UI listeners (onClick style) type of pattern then you shouldn't be doing inheritance